### PR TITLE
Show associated node labels in table view

### PR DESCRIPTION
**Pull request by mPokornyETM**
See [JENKINS-43727](https://issues.jenkins-ci.org/browse/JENKINS-43727).

I found in the last release new issue. The labels in the computer page are no more listed correctly. There are no spaces between labels and the combo box is also not really accessible. See the red marked area.
This PR partly implement changes like in 43727 describes and I think the issue 43727 can be closed.
This PR will fix this, so the labels are listed in sortable table with other helpful information.
image

### Proposed changelog entries

* Show computer labels in sortable table.

### Proposed upgrade guidelines

N/A

### Submitter checklist

- [x ] (If applicable) Jira issue is well described
- [ ] Changelog entries and upgrade guidelines are appropriate for the audience affected by the change (users or developer, depending on the change) and are in the imperative mood. [Examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)
  * Fill-in the `Proposed changelog entries` section only if there are breaking changes or other changes which may require extra steps from users during the upgrade
- [ ] Appropriate autotests or explanation to why this change has no tests
- [ ] New public classes, fields, and methods are annotated with `@Restricted` or have `@since TODO` Javadoc, as appropriate. 
- [ ] For dependency updates: links to external changelogs and, if possible, full diffs

### Desired reviewers

@mention

### Maintainer checklist

Before the changes are marked as `ready-for-merge`: 

- [ ] There are at least 2 approvals for the pull request and no outstanding requests for change
- [ ] Conversations in the pull request are over OR it is explicit that a reviewer does not block the change
- [ ] Changelog entries in the PR title and/or `Proposed changelog entries` are accurate, human-readable, and in the imperative mood
- [ ] Proper changelog labels are set so that the changelog can be generated automatically
- [ ] If the change needs additional upgrade steps from users, `upgrade-guide-needed` label is set and there is a `Proposed upgrade guidelines` section in the PR title. ([example](https://github.com/jenkinsci/jenkins/pull/4387))
- [ ] If it would make sense to backport the change to LTS, a Jira issue must exist, be a _Bug_ or _Improvement_, and be labeled as `lts-candidate` to be considered (see [query](https://issues.jenkins-ci.org/issues/?filter=12146)).

### CODE REVIEW

File: core/src/main/resources/hudson/model/Computer/index.jelly [hunk 1]
```diff
@@ -74,16 +74,46 @@ THE SOFTWARE.
...
-          </j:forEach>
+          <table id="labels" class="jenkins-table sortable">
+            <thead>
+              <tr>
+                <th class="jenkins-table__cell--tight">S</th>
+                <th initialSortDir="down">${%Name}</th>
+                <th>${%Description}</th>
+                <th align="right" tooltip="Number of associated nodes">${%Associated nodes}</th>
+                <th align="right" tooltip="Number of total Executors that belong to this label">${%Configured executors}</th>
+                <th align="right" tooltip="Number of total Executkors that belong to this label that are functioning">${%Total configured}</th>
```
Comment (NotMyFault) [inline core/src/main/resources/hudson/model/Computer/index.jelly:88]:
```suggestion
                <th align="right" tooltip="Number of total Executors that belong to this label that are functioning">${%Total configured}</th>
```

File: core/src/main/resources/hudson/model/Computer/index.jelly [hunk 2]
```diff
@@ -74,16 +74,46 @@ THE SOFTWARE.
...
+              </tr>
+            </thead>
+            <tbody>
+            <j:forEach var="entry" items="${it.node.labelCloud}">
+              <!-- Skip the label for this node -->
+              <j:if test="${entry.item!=it.node.selfLabel}">
+                <tr id="label_${entry.item.name}">
+                  <td data="${entry.item.icon}" class="jenkins-table__cell--tight jenkins-table__icon">
+                    <div class="jenkins-table__cell__button-wrapper">
+                      <l:icon src="${entry.item.iconClassName}" alt="${entry.item.iconAltText}"/>
```
Comment (NotMyFault) [inline core/src/main/resources/hudson/model/Computer/index.jelly:100]: Not every label does necessarily have an icon. The Platformlabeler plugin for example adds labels without icons, which would reference to a non-existing image in this case.

**Review by NotMyFault — COMMENTED** [review]
Also, thank you and congrats for raising your first pull request 🎉

File: core/src/main/java/hudson/model/Label.java [hunk 3]
```diff
@@ -353,6 +353,46 @@ public boolean isOffline() {
...
+     */
+    @Exported
+    public String getIconClassName() {
+        // All nodes are offline
+        if (isOffline()) return "symbol-computer-offline";
+        return "symbol-computer";
+    }
+
+    /** Returns icon as alt text. 
+      Is used in labels table to enable sorting by state
```
Comment (NotMyFault) [inline core/src/main/java/hudson/model/Label.java:388]: This line would fire a Checkstyle violation, but overall, can't we rely on the item shipped by default? If I'm not mistaken, some labels already ship the computer icon 🤔

**Comment by mPokornyETM** [comment]
Where can I found current style rules to eliminate *Checkstyle violation* issues before I commit the changes here?

**Comment by NotMyFault** [comment]
> Where can I found current style rules to eliminate _Checkstyle violation_ issues before I commit the changes here?

The [parent pom](https://github.com/jenkinsci/jenkins/blob/master/pom.xml#L497) ships the defined rules, but if you build Jenkins according to the one liner linked in the [contributing.md](https://github.com/jenkinsci/jenkins/blob/master/CONTRIBUTING.md#building-and-debugging), you can run over checkstyle violations locally.

**Comment by NotMyFault** [comment]
> I am not sure what happens. The win tests are not finished in 5h. I think it is aborted by timeout, but I have no Idea why it takes so long. Can somebody help me?

Testing core on Windows takes more than twice the time than testing on Linux, though agents timing out or being disconnected is not a merge or review blocker, as long as the actual test suite is fine.

**Comment by mPokornyETM** [comment]
> > I am not sure what happens. The win tests are not finished in 5h. I think it is aborted by timeout, but I have no Idea why it takes so long. Can somebody help me?
> 
> Testing core on Windows takes more than twice the time than testing on Linux, though agents timing out or being disconnected is not a merge or review blocker, as long as the actual test suite is fine.

great. that means it is ready to merge. thx

**Review by NotMyFault — COMMENTED** [review]
I don't think we need a symbol in the table view at all, because it mirrors the state of the node, which already is covered in the header next to the name:

The table looks fine without a symbol as well.

**Comment by mPokornyETM** [comment]
> I don't think we need a symbol in the table view at all, because it mirrors the state of the node, which already is covered in the header next to the name:  The table looks fine without a symbol as well.

I think also about that, if is it necessary. But it is little other state.  This is about online state of all associated nodes, not only about this node. That means when all associated nodes are offline this state will be also offline. The implementation for calculating was done before me, so I just show the state.
But when you want I will remove it.
Any case it will be fine, when is there some possibility to shown / hide column in the table depends on user settings. But this is other story. Is there some solution?

bad news: I see next side effect from the combo box: When you click on the label and you get listed all assigned nodes :-(
image

I will fix it in new PR. My thinking is about "new" table for nodes with filter possibility, which can be used here and in the /computer self. So it will be possible to make filters for my teams and I does not need to care about the nodes. I am just lazy ;-) and will spare my time.

**Comment by NotMyFault** [comment]
> [](https://user-images.githubusercontent.com/89339813/160546487-8708b10a-8b8b-4cfa-865b-79a49c704530.png)bad news: I see next side effect from the combo box: When you click on the label and you get listed all assigned nodes :-(
image

I don't see that being an issue in particular, because nodes can share labels. Though the overview to display that suffers from the same concerns this PR addresses.

**Comment by mPokornyETM** [comment]
> 

yes you true, I only want to fix it in another PR, that can be worked parallel and we does not wait a long time to push this changes in master. Step by step.

File: core/src/main/java/hudson/model/Label.java [hunk 4]
```diff
@@ -353,6 +353,48 @@ public boolean isOffline() {
...
+
+    /**
+     * Returns icon as alt text.
+     *
+     * Is used in labels table to enable sorting by state.
+     */
+    public String getIconAltText() {
+        // All nodes are offline
+        if (isOffline()) return "[all-offline]";
+        return "[online]";
```
Comment (NotMyFault) [inline core/src/main/java/hudson/model/Label.java:395]: New introduced user-visible messages should be setup for [localization](https://www.jenkins.io/doc/developer/internationalization/i18n-source-code/).

File: core/src/main/resources/hudson/model/Computer/index.jelly [hunk 5]
```diff
@@ -74,16 +74,46 @@ THE SOFTWARE.
...
-            <!-- Skip the label for this node -->
-            <j:if test="${entry.item!=it.node.selfLabel}">
-              <a class="${entry.className} model-link inside" href="${rootURL}/label/${entry.item.name}">${entry.item.name}</a>
-              <wbr/>
-            </j:if>
-          </j:forEach>
+          <table id="labels" class="jenkins-table sortable">
+            <thead>
+              <tr>
+                <th class="jenkins-table__cell--tight">S</th>
```
Comment (NotMyFault) [inline core/src/main/resources/hudson/model/Computer/index.jelly:83]: What does the "S" refer to? Status?

File: core/src/main/resources/hudson/model/Computer/index.jelly [hunk 6]
```diff
@@ -74,16 +74,46 @@ THE SOFTWARE.
...
+                <th initialSortDir="down">${%Name}</th>
+                <th>${%Description}</th>
+                <th align="right" tooltip="Number of associated nodes">${%Associated nodes}</th>
+                <th align="right" tooltip="Number of total executors that belong to this label">${%Configured executors}</th>
+                <th align="right" tooltip="Number of total executors that belong to this label that are functioning">${%Total configured}</th>
+                <th align="right" tooltip="Number of busy executors that are carrying out some work right now">${%Busy executors}</th>
+                <th align="right" tooltip="Number of idle executors that can start working immediately">${%Idle executors}</th>
```
Comment (NotMyFault) [inline core/src/main/resources/hudson/model/Computer/index.jelly:90]: These should be [localizable](https://www.jenkins.io/doc/developer/internationalization/i18n-jelly-views/) as well.

File: core/src/main/java/hudson/model/Label.java [hunk 7]
```diff
@@ -353,6 +353,48 @@ public boolean isOffline() {
         return true;
     }
 
+    /**
+     * Returns the icon for this label depends on state of all associated computers.
+     *
+     * see {@link #isOffline()}
+     * see {@link #getIconClassName()}
```
Comment (NotMyFault) [inline core/src/main/java/hudson/model/Label.java:360]:
```suggestion
     * @see #isOffline()
     * @see #getIconClassName()
```

File: core/src/main/java/hudson/model/Label.java [hunk 8]
```diff
@@ -353,6 +353,48 @@ public boolean isOffline() {
...
+        if (isOffline()) return "symbol-computer-offline";
+        return "symbol-computer";
+    }
+
+    /**
+     * Returns the class name that will be used to lookup the icon.
+     *
+     * This class name will be added as a class tag to the html img tags where the icon should
+     * show up followed by a size specifier given by {@link Icon#toNormalizedIconSizeClass(String)}
+     * The conversion of class tag to src tag is registered through {@link IconSet#addIcon(Icon)}
```
[the remark this answers is not in the document]
Comment (NotMyFault) [reply core/src/main/java/hudson/model/Label.java:374]: You marked this as resolved but you didn't address it? See my suggestion below.

File: core/src/main/java/hudson/model/Label.java [hunk 9]
```diff
@@ -353,6 +353,48 @@ public boolean isOffline() {
...
```
Comment (NotMyFault) [inline core/src/main/java/hudson/model/Label.java:374]:
```suggestion
     * show up followed by a size specifier given by {@link org.jenkins.ui.icon.Icon#toNormalizedIconSizeClass(String)}
     * The conversion of class tag to src tag is registered through {@link org.jenkins.ui.icon.IconSet#addIcon(org.jenkins.ui.icon.Icon)}
```
Maybe this makes it more clear what I was referring to.

File: core/src/main/java/hudson/model/Label.java [hunk 10]
```diff
@@ -353,6 +353,48 @@ public boolean isOffline() {
...
+     *
+     * It is both the recommended and default implementation to serve different icons based on {@link #isOffline}
+     *
+     * {@link #getIcon()}
```
Comment (NotMyFault) [inline core/src/main/java/hudson/model/Label.java:378]:
```suggestion
     * @see #getIcon()
```

File: core/src/main/resources/hudson/model/Computer/index.properties [hunk 11]
```diff
@@ -23,5 +23,22 @@
 submit.temporarilyOffline=Bring this node back online
 submit.not.temporarilyOffline=Mark this node temporarily offline
 submit.updateOfflineCause=Update offline reason
-title.projects_tied_on=Projects tied to {0}
+title.projects_tied_on=Projects tied to "{0}":
 title.no_manual_launch=This node\u2019s availability policy is: \u201c{0}\u201d Currently, this mandates that the node be offline.
+
+nodes.column.state.label=Status
+nodes.column.state.description=Cumulative status of nodes. The label is offline when all associated nodes are offline
```
Comment (NotMyFault) [inline core/src/main/resources/hudson/model/Computer/index.properties:30]:
```suggestion
nodes.column.state.description=Cumulative status of nodes. The label is offline when all associated nodes are offline.
```

File: core/src/main/resources/hudson/model/Computer/index_de.properties [hunk 12]
```diff
@@ -5,6 +5,23 @@ submit.not.temporarilyOffline=Knoten tempor\u00E4r abschalten
 submit.updateOfflineCause=Offline Grund aktualisieren
 title.no_manual_launch=Dieser Knoten verwendet die Verf\u00FCgbarkeitsregel \u201E{0}\u201C. \
   Dies bedeutet momentan, dass der Knoten offline ist.
-title.projects_tied_on=Projekte, die an {0} gebunden sind 
+title.projects_tied_on=Projekte, die an "{0}" gebunden sind:
 None=Keine
 Labels=Labels
+
+nodes.column.state.label=Status
+nodes.column.state.description=Kumulativer Knotenstatus, die zu diesem Label zugewiesen sind. Das Label ist offline, wenn alle zugeordneten Knoten offline sind
```
Comment (NotMyFault) [inline core/src/main/resources/hudson/model/Computer/index_de.properties:13]:
```suggestion
nodes.column.state.description=Kumulativer Knotenstatus. Das Label ist offline, wenn alle zugeordneten Knoten offline sind.
```

File: core/src/main/resources/hudson/model/Computer/index_de.properties [hunk 13]
```diff
@@ -5,6 +5,23 @@ submit.not.temporarilyOffline=Knoten tempor\u00E4r abschalten
...
+nodes.column.name.label=Name
+nodes.column.name.description=Label name
```
Comment (NotMyFault) [inline core/src/main/resources/hudson/model/Computer/index_de.properties:15]:
```suggestion
nodes.column.name.description=Label-Name
```

File: core/src/main/resources/hudson/model/Computer/index_de.properties [hunk 14]
```diff
@@ -5,6 +5,23 @@ submit.not.temporarilyOffline=Knoten tempor\u00E4r abschalten
...
+nodes.column.description.label=Beschreibung
+nodes.column.description.description=Label Beschreibung
```
Comment (NotMyFault) [inline core/src/main/resources/hudson/model/Computer/index_de.properties:17]:
```suggestion
nodes.column.description.description=Label-Beschreibung
```

File: core/src/main/resources/hudson/model/Computer/index_de.properties [hunk 15]
```diff
@@ -5,6 +5,23 @@ submit.not.temporarilyOffline=Knoten tempor\u00E4r abschalten
...
+nodes.column.noOfNodes.label=Zugeordnete Knoten
+nodes.column.noOfNodes.description=Anzahl zugeordneter Knoten
+nodes.column.noOfExecutors.label=Build-Prozessoren
+nodes.column.noOfExecutors.description=Gesamtzahl konfigurierter Build-Prozessoren, die zu diesem Label zuegeordnet sind
```
Comment (NotMyFault) [inline core/src/main/resources/hudson/model/Computer/index_de.properties:21]:
```suggestion
nodes.column.noOfExecutors.description=Gesamtzahl konfigurierter Build-Prozessoren, die diesem Label zugeordnet sind
```

File: core/src/main/resources/hudson/model/Computer/index_de.properties [hunk 16]
```diff
@@ -5,6 +5,23 @@ submit.not.temporarilyOffline=Knoten tempor\u00E4r abschalten
...
+nodes.column.noOfExecutorsOnline.label=Online Build-Prozessoren
+nodes.column.noOfExecutorsOnline.description=Gesamtzahl online Build-Prozessoren, die zu diesem Label zuegeordnet sind
```
Comment (NotMyFault) [inline core/src/main/resources/hudson/model/Computer/index_de.properties:23]:
```suggestion
nodes.column.noOfExecutorsOnline.description=Gesamtzahl der Build-Prozessoren, die online und diesem Label zugeordnet sind
```

File: core/src/main/resources/hudson/model/Computer/index_de.properties [hunk 17]
```diff
@@ -5,6 +5,23 @@ submit.not.temporarilyOffline=Knoten tempor\u00E4r abschalten
...
+nodes.column.noOfExecutorsBusy.label=Besch\u00e4ftigte Build-Prozessoren
+nodes.column.noOfExecutorsBusy.description=Gesamtzahl besch\u00e4ftigter Build-Prozessoren, die zu diesem Label zuegeordnet sind
```
Comment (NotMyFault) [inline core/src/main/resources/hudson/model/Computer/index_de.properties:25]:
```suggestion
nodes.column.noOfExecutorsBusy.description=Gesamtzahl besch\u00e4ftigter Build-Prozessoren, die diesem Label zugeordnet sind
```

File: core/src/main/resources/hudson/model/Computer/index_de.properties [hunk 18]
```diff
@@ -5,6 +5,23 @@ submit.not.temporarilyOffline=Knoten tempor\u00E4r abschalten
...
+nodes.column.noOfExecutorsIdle.label=Inaktive Build-Prozessoren
+nodes.column.noOfExecutorsIdle.description=Gesamtzahl inaktiver Build-Prozessoren, die zu diesem Label zuegeordnet sind
```
Comment (NotMyFault) [inline core/src/main/resources/hudson/model/Computer/index_de.properties:27]:
```suggestion
nodes.column.noOfExecutorsIdle.description=Gesamtzahl inaktiver Build-Prozessoren, die diesem Label zugeordnet sind
```

**Review by NotMyFault — APPROVED** [review]
Looks good so far.

**Comment by mPokornyETM** [comment]
@NotMyFault  What shall I do to complete this PR ?

**Comment by NotMyFault** [comment]
> @NotMyFault What shall I do to complete this PR ?

I'm not a core maintainer, I can't "process" your PR in any means but reviewing it, however, we can ask a @jenkinsci/core-pr-reviewers if they'd want to take a look over the PR.

**Review by daniel-beck — COMMENTED** [review]
It is unclear to me what problem this is trying to solve. This does not actually address JENKINS-43727 as I understand it, which is about getting labels in a computer list, not a labels list in a computer. Could you clarify what motivates this?

Also, I commented in Jira (second comment) about potential scalability issues, which this seems like it would run into.

**Comment by mPokornyETM** [comment]
> It is unclear to me what problem this is trying to solve. This does not actually address JENKINS-43727 as I understand it, which is about getting labels in a computer list, not a labels list in a computer. Could you clarify what motivates this?
> 
> Also, I commented in Jira (second comment) about potential scalability issues, which this seems like it would run into.

Maybe I don't understand the description in issue correctly.   I can also build Page with all labels.  It might look like nodes overview (/computer page). And should be opened also from jenkins management page. Whe you agree I will comment in issue self and I can implement it.

Anyway this pr fix the spaces between labels.  When you has more than 1 label configured by node. There are no spaces. Which is very hard to read. You ca see that in my screenshot.

**Comment by mPokornyETM** [comment]
About the @Exported. I am new in Java. Just copy & paste from other source and adapted it for my use case.  I will remove it after my vacation, probably next week.

**Comment by daniel-beck** [comment]
> Anyway this pr fix the spaces between labels. When you has more than 1 label configured by node. There are no spaces. Which is very hard to read. You ca see that in my screenshot.

Kind of, but this needs to be solved through XSS with the `model-link`, otherwise plugins like Credentials are still unaddressed. In this way, core acts both as framework and implementation, and a better fix would be through the implementation. I don't think we need to replace the sequence of label links with a table just so they don't overlap, otherwise the messaging is that every plugin with sequential model links will need to have a similar change applied.

**Comment by mPokornyETM** [comment]
> 

That means you prefer the "old" list style with fixed spacing. From my POV I prefer the table view. The computer page has enough space and you can see in one overview all important information. This was a longer time a wish from my team members.  Therfore the fix in new style. Please let me time to read something about the model-link and I will try it to fix in other way. I am back from my vacation next week. We can discuss which solution is better. 
When you agree I convert this pr to draft now.
About JENKINS-43727. I change it locally and this was similar issue for me,  so I have just linked it. You true it is misleading from my side, I didn't understand it correctly.  Shall I create new issue in jira?

**Comment by mwinter69** [comment]
Personally I agree with @daniel-beck that I prefer the existing sequence of label links over the table view.
The table adds information that is not only related to the agent we look at.
Also when e.g. using the platformlabeler plugin the list of labels will get quite long.
The label-linked-jobs already has an overview of all defined labels, but could use a better UI with a table view to show the things.

**Comment by mPokornyETM** [comment]
The issue is fixed now. For translation I will create new PR later. Therefore is this one obsolete and will be closed now. Thx for reviews.

**Pull request closed without merging**